### PR TITLE
Add compiler messages for primary constructors

### PR DESCRIPTION
**Pull request by BillWagner**
Add diagnostic messages for compiler messages related to constructor declarations.

Contributes to #34529 

Commit by commit:

1. Create stubs for primary constructor diagnostics.
1. Create content for primary constructors.
1. Migrate content for existing errors and warnings related to constructor declarations. Add stubs for undocumented messages.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/compiler-messages/constructor-errors.md](https://github.com/dotnet/docs/blob/9c0da1e2717e7180277867752991f5c2c1841edd/docs/csharp/language-reference/compiler-messages/constructor-errors.md) | [docs/csharp/language-reference/compiler-messages/constructor-errors](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/constructor-errors?branch=pr-en-us-35324) |
| [docs/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error.md](https://github.com/dotnet/docs/blob/9c0da1e2717e7180277867752991f5c2c1841edd/docs/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error.md) | [Sorry, we don't have specifics on this C# error](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error?branch=pr-en-us-35324) |

### CODE REVIEW

**Comment by BillWagner** [comment]
Internal previews before final commit that fixes links:

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/compiler-messages/constructor-errors.md](https://github.com/dotnet/docs/blob/78b255e6cd7ef207c64bc451a7b783db05ba227b/docs/csharp/language-reference/compiler-messages/constructor-errors.md) | [docs/csharp/language-reference/compiler-messages/constructor-errors](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/constructor-errors?branch=pr-en-us-35324) |
| [docs/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error.md](https://github.com/dotnet/docs/blob/78b255e6cd7ef207c64bc451a7b783db05ba227b/docs/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error.md) | [Sorry, we don't have specifics on this C# error](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/sorry-we-don-t-have-specifics-on-this-csharp-error?branch=pr-en-us-35324) |

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+- [**CS8910**](#records-and-copy-constructors) - *The primary constructor conflicts with the synthesized copy constructor.*
+- [**CS8958**](#constructors-in-struct-types) - *The parameterless struct constructor must be 'public'.*
+- [**CS8982**](#constructors-in-struct-types) - *A constructor declared in a 'struct' with parameter list must have a 'this' initializer that calls the primary constructor or an explicitly declared constructor.*
+- [**CS8983**](#constructors-in-struct-types) - *A 'struct' with field initializers must include an explicitly declared constructor.*
+- [**CS9105**](#primary-constructor-syntax) - *Cannot use primary constructor parameter in this context.*
+- [**CS9106**](#primary-constructor-syntax) - *Identifier is ambiguous between type and parameter in this context.*
+class as well.*
+- [**CS9108**](#primary-constructor-syntax) - *Cannot use parameter that has ref-like type inside an anonymous method, lambda expression, query expression, or local function.*
+- [**CS9109**](#primary-constructor-syntax) - *Cannot use ref, out, or in primary constructor parameter inside an instance member.*
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:117]:
```suggestion
- [**CS9109**](#primary-constructor-syntax) - *Cannot use `ref`, `out`, or `in` primary constructor parameter inside an instance member.*
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+- [**CS9114**](#primary-constructor-syntax) - *A primary constructor parameter of a readonly type cannot be assigned to (except in init-only setter of the type or a variable initializer).*
+- [**CS9115**](#primary-constructor-syntax) - *A primary constructor parameter of a readonly type cannot be returned by writable reference.*
+- [**CS9116**](#primary-constructor-syntax) - *A primary constructor parameter of a readonly type cannot be used as a ref or out value (except in init-only setter of the type or a variable initializer).*
+- [**CS9117**](#primary-constructor-syntax) - *Members of primary constructor parameter of a readonly type cannot be modified (except in init-only setter of the type or a variable initializer).*
+- [**CS9118**](#primary-constructor-syntax) - *Members of primary constructor parameter of a readonly type cannot be returned by writable reference.*
+- [**CS9119**](#primary-constructor-syntax) - *Members of primary constructor parameter' of a readonly type cannot be used as a ref or out value (except in init-only setter of the type or a variable initializer).*
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:126]:
```suggestion
- [**CS9119**](#primary-constructor-syntax) - *Members of primary constructor parameter of a readonly type cannot be used as a `ref` or `out` value (except in init-only setter of the type or a variable initializer).*
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+- [**CS9113**](#primary-constructor-syntax) - *Parameter is unread.*
+
+## Static constructors
+
+- **CS0514** - *Static constructor cannot have an explicit 'this' or 'base' constructor call.*
+- **CS0515** - *Access modifiers are not allowed on static constructors.*
+
+You can write at most one static constructor for a type. The declaration of a static constructor must obey the following rules:
+
+- The static constructor has the `static` modifier, but no other modifiers such as `public`, `protected`, `private` or `internal`.
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:144]:
```suggestion
- The static constructor has the `static` modifier, but no other modifiers, such as `public`, `protected`, `private`, or `internal`.
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+- The static constructor must be a parameterless constructor.
+- The static constructor must not call `base()`, or `this()`. If the base class includes a static constructor, the runtime calls it automatically.
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:146]:
```suggestion
- The static constructor must not call `base()` or `this()`. If the base class includes a static constructor, the runtime calls it automatically.
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+- **CS0526** - *Interfaces cannot contain constructors.*
+- **CS0710** - *Static classes cannot have instance constructors.*
+- **CS8054** - *Enums cannot contain explicit parameterless constructors.*
+- **CS8358** - *Cannot use attribute constructor because it has 'in' parameters.*
+- **CS8091** - *A constructor cannot be extern and have a constructor initializer.*
+
+Constructors are allowed only in `class` and `struct` types, including `record class` and `record struct` types. You can't define them in `enum` or `interface` types. Furthermore, [attribute](../attributes/general.md) class types can't declare `in` parameters. Instead, pass parameters by value.
+
+You can declare `extern` constructors, however you can't use `base()`, or `this()` constructor calls to call another constructor from a constructor declared `extern`.
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:158]:
```suggestion
You can declare `extern` constructors, however you can't use `base()` or `this()` constructor calls to call another constructor from a constructor declared `extern`.
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+When a constructor is marked `extern`, the compiler can't guarantee the constructor exists. Therefore, the compiler generates this warning.
+
+## Constructors in struct types
+
+- **CS0568** - *Structs cannot contain explicit parameterless constructors.*
+- **CS8958** - *The parameterless struct constructor must be 'public'.*
+- **CS8982** - *A constructor declared in a 'struct' with parameter list must have a 'this' initializer that calls the primary constructor or an explicitly declared constructor.*
+- **CS8983** - *A 'struct' with field initializers must include an explicitly declared constructor.*
+
+Recent features in C# remove earlier restrictions to `struct` types. **CS0568** is generated when you declare a parameterless constructor and you're using C# 9 or earlier. Once you're using C#10, you can declare an explicit parameterless constructor. That explicit parameterless constructor must be `public`. If your `struct` declares any [field initializers](../../programming-guide/classes-and-structs/fields.md), you must also declare an explicit constructor. This constructor can be a parameterless constructor with an empty body.
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:173]:
```suggestion
Recent features in C# remove earlier restrictions to `struct` types. **CS0568** is generated when you declare a parameterless constructor and you're using C# 9 or earlier. Once you're using C# 10, you can declare an explicit parameterless constructor. That explicit parameterless constructor must be `public`. If your `struct` declares any [field initializers](../../programming-guide/classes-and-structs/fields.md), you must also declare an explicit constructor. This constructor can be a parameterless constructor with an empty body.
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+When a `struct` type declares a primary constructor, including `record struct` types, all other constructors except a parameterless constructor must call the primary constructor, or another explicitly declared constructor using `this()`.
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:175]:
```suggestion
When a `struct` type declares a primary constructor, including `record struct` types, all other constructors except a parameterless constructor must call the primary constructor or another explicitly declared constructor using `this()`.
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+## Constructor calls with `base` and `this`
+
+- **CS0516** - *Constructor can not call itself.*
+- **CS0517** - *Class has no base class and cannot call a base constructor.*
+- **CS0522** - *Structs cannot call base class constructors.*
+- **CS0768** - *Constructor cannot call itself through another constructor.*
+
+You can use `base()` and `this()` to have one constructor call another in the same type or the base type. Calling constructors can minimize duplicated constructor logic. You must follow these rules when calling another constructor using `this()` or `base()`:
+
+- You can't call the same constructor, either directly, or indirectly through another constructor. For example, the following code is illegal:
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:186]:
```suggestion
- A constructor can't call itself either directly or indirectly through another constructor. For example, the following code is illegal:
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+
+## Primary constructor syntax
+
+The compiler emits the following errors when a primary constructor violates one or more rules on primary constructors for classes and structs:
+
+- **CS8862** - *A constructor declared in a type with parameter list must have 'this' constructor initializer.*
+- **CS9105** - *Cannot use primary constructor parameter in this context.*
+- **CS9106** - *Identifier is ambiguous between type and parameter in this context.*
+- **CS9108** - *Cannot use parameter that has ref-like type inside an anonymous method, lambda expression, query expression, or local function.*
+- **CS9109** - *Cannot use ref, out, or in primary constructor parameter inside an instance member.*
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:232]:
```suggestion
- **CS9109** - *Cannot use `ref`, `out`, or `in` primary constructor parameter inside an instance member.*
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+- **CS9112** - *Anonymous methods, lambda expressions, query expressions, and local functions inside a struct cannot access primary constructor parameter also used inside an instance member.*
+- **CS9114** - *A primary constructor parameter of a readonly type cannot be assigned to (except in init-only setter of the type or a variable initializer).*
+- **CS9115** - *A primary constructor parameter of a readonly type cannot be returned by writable reference.*
+- **CS9116** - *A primary constructor parameter of a readonly type cannot be used as a ref or out value (except in init-only setter of the type or a variable initializer).*
+- **CS9117** - *Members of primary constructor parameter of a readonly type cannot be modified (except in init-only setter of the type or a variable initializer).*
+- **CS9118** - *Members of primary constructor parameter of a readonly type cannot be returned by writable reference.*
+- **CS9119** - *Members of primary constructor parameter' of a readonly type cannot be used as a ref or out value (except in init-only setter of the type or a variable initializer).*
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:241]:
```suggestion
- **CS9119** - *Members of primary constructor parameter of a readonly type cannot be used as a `ref` or `out` value (except in init-only setter of the type or a variable initializer).*
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+- **CS9108** - *Cannot use parameter that has ref-like type inside an anonymous method, lambda expression, query expression, or local function.*
+- **CS9109** - *Cannot use ref, out, or in primary constructor parameter inside an instance member.*
+- **CS9110** - *Cannot use primary constructor parameter that has ref-like type inside an instance member.*
+- **CS9111** - *Anonymous methods, lambda expressions, query expressions, and local functions inside an instance member of a struct cannot access primary constructor parameter.*
+- **CS9112** - *Anonymous methods, lambda expressions, query expressions, and local functions inside a struct cannot access primary constructor parameter also used inside an instance member.*
+- **CS9114** - *A primary constructor parameter of a readonly type cannot be assigned to (except in init-only setter of the type or a variable initializer).*
+- **CS9115** - *A primary constructor parameter of a readonly type cannot be returned by writable reference.*
+- **CS9116** - *A primary constructor parameter of a readonly type cannot be used as a ref or out value (except in init-only setter of the type or a variable initializer).*
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:238]:
```suggestion
- **CS9116** - *A primary constructor parameter of a readonly type cannot be used as a `ref` or `out` value (except in init-only setter of the type or a variable initializer).*
```

File: docs/csharp/language-reference/compiler-messages/constructor-errors.md [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+- **CS9121** - *Struct primary constructor parameter of type causes a cycle in the struct layout.*
+- **CS9122** - *Unexpected parameter list.*
+
+Primary constructor parameters are in scope in the body of that type. The compiler can synthesize a field that stores the parameter for use in members or in field initializers. Because a primary constructor parameter may be copied to a field, the following restrictions apply:
+
+- Primary constructors can be declared on `struct` and `class` types, but not on `interface` types.
+- Primary constructor parameters can't be used in a `base()` constructor call except as part of the primary constructor.
+- Primary constructor parameters of `ref struct` type can't be accessed in lambda expressions, query expressions, or local functions.
+- If the type isn't a `ref struct`, `ref struct` parameters can't be accessed in instance members.
+- In a `ref struct` type, primary constructor parameters with the `in`, `ref` or `out` modifiers can't be used in any instance methods, or property accessors.
```
Comment (gewarren) [inline docs/csharp/language-reference/compiler-messages/constructor-errors.md:252]:
```suggestion
- In a `ref struct` type, primary constructor parameters with the `in`, `ref`, or `out` modifiers can't be used in any instance methods or property accessors.
```

**Review by gewarren — APPROVED** [review]
LGTM and I left some suggestions to consider. Also are you concerned about the error messages being localized?

**Pull request merged**